lib/propolis/src/hw/qemu/pvpanic.rs

```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering::Relaxed},
+    Arc,
+};
+
+use crate::common::*;
+use crate::pio::{PioBus, PioFn};
+
+pub struct QemuPioPvpanic {
+    counts: Arc<PanicCounts>,
+}
+
+#[derive(Debug, Default)]
+pub struct PanicCounts {
+    host_handled: AtomicUsize,
+    guest_handled: AtomicUsize,
+}
+
+/// Indicates that a guest panic has happened and should be processed by the
+/// host
+const HOST_HANDLED: u8 = 0b01;
+/// Indicates a guest panic has happened and will be handled by the guest; the
+/// host should record it or report it, but should not affect the execution of
+/// the guest.
+const GUEST_HANDLED: u8 = 0b10;
+
+impl QemuPioPvpanic {
+    const IOPORT: u16 = 0x505;
+
+    #[must_use]
+    pub fn create(pio: &PioBus, counts: Arc<PanicCounts>) -> Arc<Self> {
+        let this = Arc::new(Self { counts });
+
+        let piodev = this.clone();
+        let piofn = Arc::new(move |_port: u16, rwo: RWOp| piodev.pio_rw(rwo))
+            as Arc<PioFn>;
+        pio.register(Self::IOPORT, 1, piofn).unwrap();
```

Comment:
The other qemu devices set a bad example, and this is not (yet) well documented, but we should try to split up the device creation from its attachment to any resources.  See something like `LpcUart` or the ps/2 controller for an example.

Reply:
Ah, thanks for the advice, I'll go ahead and change that!